# DTSC-580: Data Manipulation
## Assignment: Scooby-Doo

### Name: Enter Name Before Submitting

## Copyright & Academic Integrity Notice
<span style="color:red">This material is for enrolled students' academic use only and protected under U.S. Copyright Laws. This content must not be shared outside the confines of this course, in line with Eastern University's academic integrity policies. Unauthorized reproduction, distribution, or transmission of this material, including but not limited to posting on third-party platforms like GitHub, is strictly prohibited and may lead to disciplinary action. You may not alter or remove any copyright or other notice from copies of any content taken from BrightSpace or Eastern University’s website.</span>
 
<span style="color:red">© Copyright Notice 2025, Eastern University - All Rights Reserved.</span> 

## Overview

In this assignment, you will test the skills that you have learned during this course to manipulate the provided data to answer questions about the TV show Scooby-Doo.  If you are not familiar with this show, more information can be found on [Scoobypedia](https://scoobydoo.fandom.com/wiki/Scoobypedia).

> Scoobypedia is an encyclopedia on the hit television series Scooby-Doo which has been airing for over 50 years!
> 
> The show follows the iconic mystery solving detectives, known as Mystery Inc., as they set out in the Mystery Machine to solve crime and unmask criminals, bent on revenge or committing criminal acts for their own personal gain.
> 
> Titular character, Scooby, is accompanied by his best pal Shaggy as both vie for Scooby Snacks on their adventures! Velma brings her laser-focused intellect and initiative to collecting clues and solving mysteries. Fred is the team's leader who devises elaborate traps to catch the bad guys, while Daphne is bold and empathetic and keeps the Gang together.

Please note that your notebook should be named `scooby` when submitting to CodeGrade for the automatic grading to work properly.

## About Dataset

This data comes from [Kaggle](https://www.kaggle.com/datasets/williamschooleman/scoobydoo-complete) thanks to manual data aggregation by Kaggle user plummye. We have included a PDF from the dataset's author that provides detailed information on how the data was collected, additional details about the features, and how the data was analyzed to identify features potentially important to the IMDB score.

## Data Dictionary

### scoobydoo_episodes
| variable | description |
| ----------| ---------- |
| index	| index based on Scoobypedia |
| series_name | Name of the series in which the episode takes place or in movies' cases the Scoobypedia's grouping classification |
| network | Network the TV series takes place in, if it is a movie will use similar grouping as series.name variable |
| season | Season of TV Series, if not TV Series will default to the format |
| title | Title of Episode/Movie |
| imdb | Score on IMDB (NULL if recently aired) |
| engagement | Number of reviews on IMDB (NULL if very recently aired) |
| date_aired | Dated aired in US |
| run_time | Run time in min |
| format | Type |

### scoobydoo_monsters
| variable | description |
| ---------- | ---------|
|index | index based on Scoobypedia |
|monster_name | name of monster |
|monster_gender | monster gender |
|monster_type | monster type |
|monster_subtype | monster subtype |
|monster_species | monster_species |
|monster_real | was monster real |
|monster_amount | monster amount |
|caught_fred | caught by Fred |
|caught_daphnie | caught by Daphnie |
|caught_velma | caught by Velma |
|caught_shaggy | caught by Shaggy |
|caught_scooby | caught by Scooby |
|captured_fred | captured Fred |
|captured_daphnie |	captured Daphnie |
|captured_velma  | captured Velma |
|captured_shaggy | captured Shaggy |
|captured_scooby | captured Scooby |
|unmask_fred | unmask by fred |
|unmask_daphnie | unmask by Daphnie |
|unmask_velma | unmask by Velma |
|unmask_shaggy | unmask by Shaggy |
|unmask_scooby | unmask by Scooby |
|snack_fred | snack offered by Fred |
|snack_daphnie | snack offered by Daphnie |
|snack_velma | snack offered by Velma |
|snack_shaggy | snack offered by Shaggy |
|snack_scooby | snack offered by Scooby |
|unmask_other |	unmask by other |
|caught_other | caught by other |
|caught_not | not caught |
|trap_work_first | trap work first |
|setting_terrain | setting type of terrain |
|setting_country_state | setting country state |
|suspects_amount | suspects amount |
|non_suspect | non suspect |
|arrested | arrested |
|culprit_name |	culprit name |
|culprit_gender | culprit gender |
|culprit_amount | culprit amount |
|motive | motive |
|if_it_wasnt_for | phrase at the end of show, ie "if it wasnt for ..." |
|and_that | and that |
|door_gag | door gag |
|number_of_snacks | number of snacks |
|split_up |	split up |
|another_mystery | another mystery |
|set_a_trap | set a trap |
|jeepers | Times "jeepers" said |
|jinkies | Times "jinkies" said |
|my_glasses	| Times "my glasses" said |
|just_about_wrapped_up | Times "just about wrapped up" said |
|zoinks	| Times "zoinks"said |
|groovy	| Times "groovy" said |
|scooby_doo_where_are_you | Times "scooby doo where are you" said |
|rooby_rooby_roo | Times "rooby_rooby_roo" said |
|batman | batman in episode |
|scooby_dum | scooby_dum in episode |
|scrappy_doo | scrappy_doo in episode |
|hex_girls | hex_girls in episode |
|blue_falcon | blue_falcon in episode |
|fred_va | Fred voice actor |
|daphnie_va | Daphnie voice actor |
|velma_va | Velma voice actor |
|shaggy_va | Shaggy voice actor |
|scooby_va | Scooby voice actor |

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this.


## Initial Import & Exploration

In [640]:
# standard imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings("ignore")

Start by importing your two data files. Take time to thoroughly explore the data. Investing effort in understanding the dataset upfront typically makes data manipulation tasks more efficient and manageable during a project.

In [642]:
### Import Data Files ###
episodes = pd.read_csv('scoobydoo_episodes.csv')
monsters = pd.read_csv('scoobydoo_monsters.csv')

In [643]:
### Explore Data ###
monsters.info

<bound method DataFrame.info of      Index                                 Monster Name       Monster Gender  \
0        1                                 Black Knight                 Male   
1        2                       Ghost of Cptn. Cuttler                 Male   
2        3                                      Phantom                 Male   
3        4                                  Miner 49'er                 Male   
4        5                          Indian Witch Doctor               Female   
..     ...                                          ...                  ...   
603    604  Hyde Monster,Mummy,Dinosaur,Giant Manta Ray  Male,Male,Male,Male   
604    605         Alpha Jackal-Lantern,Jackal-Lanterns            Male,Male   
605    606                   Morgan le Fay,Black Knight          Female,Male   
606    607                      Mud Men,Mud Men,Mud Men       Male,Male,Male   
607    608                                 Max Kilobyte                 Male   

       

## Exercises

**Exercise1: Merging Datasets**

Merge the two datasets and name the resulting DataFrame `scoobydoo_merged`.

In [646]:
### ENTER CODE HERE ###
scoobydoo_merged =pd.merge(episodes, monsters)

**Code Check:** The shape of `scoobydoo_merged` should be `(608, 75)` after this initial merge. Note that the shape of the dataset will be different  at the end of the assignment due to some cleaning steps that follow.

In [648]:
### CODE CHECK ###
scoobydoo_merged.shape

(608, 75)

<hr/>

**Exercise2: Updating Column Names** 

Change every column name to lowercase and snake case.  This is a standard first step for some programmers as lowercase makes it easier to write and snake case makes it easier to copy multiple-word column names.

For example, `Series Name` should end up being `series_name`. It's fine if you don't, but you should be able to accomplish this in a single line of code without using a for loop.

In [650]:
### ENTER CODE HERE ###
scoobydoo_merged.columns = scoobydoo_merged.columns.str.strip().str.lower().str.replace(r"[^\w]+", "_", regex=True)

**Code Check:** Check your columns to make sure your changes worked as expected and were saved to the `scoobydoo_merged` DataFrame.

In [652]:
### CODE CHECK ###
scoobydoo_merged.head()

,index,series_name,network,season,title,imdb,engagement,date_aired,run_time,format,monster_name,monster_gender,monster_type,monster_subtype,monster_species,monster_real,monster_amount,caught_fred,caught_daphnie,caught_velma,caught_shaggy,caught_scooby,captured_fred,captured_daphnie,captured_velma,captured_shaggy,captured_scooby,unmask_fred,unmask_daphnie,unmask_velma,unmask_shaggy,unmask_scooby,snack_fred,snack_daphnie,snack_velma,snack_shaggy,snack_scooby,unmask_other,caught_other,caught_not,trap_work_first,setting_terrain,setting_country_state,suspects_amount,non_suspect,arrested,culprit_name,culprit_gender,culprit_amount,motive,if_it_wasnt_for,and_that,door_gag,number_of_snacks,split_up,another_mystery,set_a_trap,jeepers,jinkies,my_glasses,just_about_wrapped_up,zoinks,groovy,scooby_doo_where_are_you,rooby_rooby_roo,batman,scooby_dum,scrappy_doo,hex_girls,blue_falcon,fred_va,daphnie_va,velma_va,shaggy_va,scooby_va
0,1,"Scooby Doo, Where Are You!",CBS,1,What a Night for a Knight,8.1,556.0,9/13/1969,21 mins,TV Series,Black Knight,Male,Possessed Object,Suit,Object,False,1,False,False,False,True,True,False,False,False,False,False,False,False,False,False,True,True,False,False,False,False,False,False,False,NaN,Urban,United States,2,False,True,Mr. Wickles,Male,1,Theft,NaN,NaN,False,2,1.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
1,2,"Scooby Doo, Where Are You!",CBS,1,A Clue for Scooby Doo,8.1,479.0,9/20/1969,22 mins,TV Series,Ghost of Cptn. Cuttler,Male,Ghost,Suit,Human,False,1,False,False,False,True,False,True,True,True,False,False,True,False,False,False,False,False,False,True,False,False,False,False,False,False,Coast,United States,2,True,True,Cptn. Cuttler,Male,1,Theft,NaN,NaN,False,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,3.0,0.0,1.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
2,3,"Scooby Doo, Where Are You!",CBS,1,Hassle in the Castle,8.0,455.0,9/27/1969,21 mins,TV Series,Phantom,Male,Ghost,Phantom,Human,False,1,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,True,False,False,False,False,False,False,False,False,Island,United States,0,True,True,Bluestone the Great,Male,1,Treasure,NaN,NaN,False,3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,2.0,0.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
3,4,"Scooby Doo, Where Are You!",CBS,1,Mine Your Own Business,7.8,426.0,10/4/1969,21 mins,TV Series,Miner 49'er,Male,Ancient,Miner,Human,False,1,True,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,False,False,True,Cave,California,2,False,True,Hank,Male,1,Natural Resource,NaN,NaN,False,2,1.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,1.0,0.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
4,5,"Scooby Doo, Where Are You!",CBS,1,Decoy for a Dognapper,7.5,391.0,10/11/1969,21 mins,TV Series,Indian Witch Doctor,Female,Ancient,Witch Doctor,Human,False,1,False,False,False,False,True,False,False,False,False,True,False,False,False,False,True,False,True,False,False,False,False,False,False,NaN,Desert,United States,1,False,True,Buck Master,Male,1,Competition,NaN,NaN,False,2,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick


<hr/>

**Exercise3: Handle Duplicates**

1) Identify and remove any duplicate rows saving your changes back to `scoobydoo_merged`. For this assignment, assume duplicates have the same `title` and `date_aired`.
2) Reset the index for `scoobydoo_merged` ensuring that it goes from 0 to n-1.

In [654]:
### ENTER CODE HERE ###
scoobydoo_merged = scoobydoo_merged.drop_duplicates(subset = ['title', 'date_aired'], keep='first')
scoobydoo_merged = scoobydoo_merged.reset_index(drop=True)                                 

**Code Check:** Check the shape of `scoobydoo_merged`; it should now be `(603, 75)`.

In [656]:
### CODE CHECK ###
scoobydoo_merged.shape

(603, 75)

<hr/>

**Exercise4: Clean Runtime Column Part 1**

To get ready for a later exercise, we need to clean the `run_time` column. Let's start by updating all `run_time` values in `scoobydoo_merged` to floats (e.g., clean values like `21 mins` to the float `21.0`).

In [658]:
### ENTER CODE HERE ###
scoobydoo_merged['run_time']=scoobydoo_merged['run_time'].str.replace(" mins", "", regex=False).astype(float)

**Code Check:** Check that the first five values of the `run_time` column match the following output:

```
0      21.0
1      22.0
2      21.0
3      21.0
4      21.0
```

In [660]:
### CODE CHECK ###
scoobydoo_merged.head(5)

,index,series_name,network,season,title,imdb,engagement,date_aired,run_time,format,monster_name,monster_gender,monster_type,monster_subtype,monster_species,monster_real,monster_amount,caught_fred,caught_daphnie,caught_velma,caught_shaggy,caught_scooby,captured_fred,captured_daphnie,captured_velma,captured_shaggy,captured_scooby,unmask_fred,unmask_daphnie,unmask_velma,unmask_shaggy,unmask_scooby,snack_fred,snack_daphnie,snack_velma,snack_shaggy,snack_scooby,unmask_other,caught_other,caught_not,trap_work_first,setting_terrain,setting_country_state,suspects_amount,non_suspect,arrested,culprit_name,culprit_gender,culprit_amount,motive,if_it_wasnt_for,and_that,door_gag,number_of_snacks,split_up,another_mystery,set_a_trap,jeepers,jinkies,my_glasses,just_about_wrapped_up,zoinks,groovy,scooby_doo_where_are_you,rooby_rooby_roo,batman,scooby_dum,scrappy_doo,hex_girls,blue_falcon,fred_va,daphnie_va,velma_va,shaggy_va,scooby_va
0,1,"Scooby Doo, Where Are You!",CBS,1,What a Night for a Knight,8.1,556.0,9/13/1969,21.0,TV Series,Black Knight,Male,Possessed Object,Suit,Object,False,1,False,False,False,True,True,False,False,False,False,False,False,False,False,False,True,True,False,False,False,False,False,False,False,NaN,Urban,United States,2,False,True,Mr. Wickles,Male,1,Theft,NaN,NaN,False,2,1.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
1,2,"Scooby Doo, Where Are You!",CBS,1,A Clue for Scooby Doo,8.1,479.0,9/20/1969,22.0,TV Series,Ghost of Cptn. Cuttler,Male,Ghost,Suit,Human,False,1,False,False,False,True,False,True,True,True,False,False,True,False,False,False,False,False,False,True,False,False,False,False,False,False,Coast,United States,2,True,True,Cptn. Cuttler,Male,1,Theft,NaN,NaN,False,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,3.0,0.0,1.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
2,3,"Scooby Doo, Where Are You!",CBS,1,Hassle in the Castle,8.0,455.0,9/27/1969,21.0,TV Series,Phantom,Male,Ghost,Phantom,Human,False,1,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,True,False,False,False,False,False,False,False,False,Island,United States,0,True,True,Bluestone the Great,Male,1,Treasure,NaN,NaN,False,3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,2.0,0.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
3,4,"Scooby Doo, Where Are You!",CBS,1,Mine Your Own Business,7.8,426.0,10/4/1969,21.0,TV Series,Miner 49'er,Male,Ancient,Miner,Human,False,1,True,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,False,False,True,Cave,California,2,False,True,Hank,Male,1,Natural Resource,NaN,NaN,False,2,1.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,1.0,0.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick
4,5,"Scooby Doo, Where Are You!",CBS,1,Decoy for a Dognapper,7.5,391.0,10/11/1969,21.0,TV Series,Indian Witch Doctor,Female,Ancient,Witch Doctor,Human,False,1,False,False,False,False,True,False,False,False,False,True,False,False,False,False,True,False,True,False,False,False,False,False,False,NaN,Desert,United States,1,False,True,Buck Master,Male,1,Competition,NaN,NaN,False,2,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,False,False,False,False,False,Frank Welker,Stefanianna Christopherson,Nicole Jaffe,Casey Kasem,Don Messick


<hr/>

**Exercise5: Clean Runtime Column Part 2**

The `run_time` column has missing values for each format type. Impute these missing values by replacing them with the average run time for their respective format type rounded to two decimals. For example, use the average run time for the "TV Series" format (rounded to 2 decimals) to fill null values for rows with that format, use the average run time for the "TV Series (segmented)" format (rounded to 2 decimals) to fill null values for rows with that format, etc.

In [662]:
### ENTER CODE HERE ###
runtime_avg = scoobydoo_merged.groupby('format')['run_time'].transform('mean').round(2)
scoobydoo_merged['run_time'] = scoobydoo_merged['run_time'].fillna(runtime_avg)

**Code Check:** 1) Ensure the `run_time` column has no missing values. 2) To confirm you imputed the correct mean values, the sum of the column should now be `14183.73`.

In [664]:
### CODE CHECK ###
scoobydoo_merged['run_time'].sum()

14183.73

<hr/>

**Exercise6: Summarize Runtime by Format**

With the `run_time` column cleaned, analyze how the average run time differs across the various format types. Group your data to determine the average run time for each format type, rounded to two decimal places. Arrange your results in ascending order and save them as a Series named `Exercise6`.

In [666]:
### ENTER CODE HERE ###
Exercise6 = scoobydoo_merged.groupby('format')['run_time'].mean().round(2).sort_values(ascending=True)


<hr/>

**Exercise7: Clean the Network Column**

Address the inconsistencies in the network column by standardizing entries for 'Cartoon Network.' Ensure all variations of the name (e.g., spelling differences) are replaced with a consistent format: 'Cartoon Network.'"

Hint: You can reuse code from a previous assignment for this step. As your data manipulation skills grow, you'll notice recurring patterns in your tasks. Building a library of reusable code will be valuable for future projects.

In [668]:
### ENTER CODE HERE ###
network = {'Cartoon  Network': 'Cartoon Network', 
           'CartoonNetwork': 'Cartoon Network', 
           'Cartoon Ntwrk': 'Cartoon Network',
           'cartoon_network': 'Cartoon Network',
           'Cartoon Netwrk': 'Cartoon Network',
           ' Cartoon Network': 'Cartoon Network',
           'Cartoon Network ': 'Cartoon Network'
          }
scoobydoo_merged['network'] = scoobydoo_merged['network'].replace(network)

#fec["cand_nm"][123456:123461].map(parties)

**Code Check:** Display the unique network values to verify that all variations of "Cartoon Network" have been combined.

In [670]:
### CODE CHECK ###

scoobydoo_merged['network'].unique()

array(['CBS', 'ABC', 'Syndication', 'TBC', 'Cartoon Network',
       'Warner Home Video', 'Warner Bros. Picture', 'Adult Swim',
       'The WB', 'The CW', 'Boomerang'], dtype=object)

<hr/>

**Exercise8: Network with Highest IMDB Scores**

With the `network` column cleaned, analyze the average IMDB score for each network. Identify the network with the highest average score by grouping the data, calculating the averages (rounded to two decimals), and organizing the results in ascending order. Save your final output as a Series named `Exercise8`.

In [672]:
### ENTER CODE HERE ###
#Exercise6 = scoobydoo_merged.groupby('format')['run_time'].mean().round(2).sort_values(ascending=True)

Exercise8= scoobydoo_merged.groupby('network')['imdb'].mean().round(2).sort_values(ascending=True)

#Exercise8.tail()

<hr/>

**Exercise9: Who Caught the Most Monsters**

Determine which characters caught the most monsters: Fred, Daphne, Velma, Shaggy, or Scooby. Use the columns related to monster captures (refer to the data dictionary for details) to calculate the total number of captures for each character. Sort the results in ascending order and save your final output as a Series named `Exercise9`.

In [674]:
### ENTER CODE HERE ###
#caught_daphnie
#captured_scooby
capture_cols = ['caught_fred', 'caught_daphnie', 'caught_velma',
                'caught_shaggy', 'caught_scooby']
Exercise9 = scoobydoo_merged[capture_cols].sum().sort_values(ascending=True)

<hr/>

**Exercise10: Who Was Most Succcessful At Traps**

In the previous exercise, you identified who caught the most monsters. Now, refine this analysis by focusing only on episodes where the trap worked on the first try. Determine who was the most successful in catching monsters under these conditions.

Sort the results in ascending order and save the result as a Series named `Exercise10`.

Hint: You may find the code from the previous exercise helpful but will need to add an additional condition.

In [676]:
### ENTER CODE HERE ###
trap_success = scoobydoo_merged[scoobydoo_merged['trap_work_first'] == True]

Exercise10 = trap_success[capture_cols].sum().sort_values(ascending = True)

<hr/>

**Exercise11: Add Year Aired Column**

You want to determine when the show was at its peak based on total number of episodes per year. In order to get this information, you'll add a new column called `year_aired` that strips the year from the `date_aired` column.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_merged_year`. This ensures the original data remains unchanged, making CodeGrade grading easier.
2) Add a new column, `year_aired`, to the copied DataFrame. Extract the year from the `date_aired` column and assign it to this new column. Ensure that the new column is an integer data type.

In [678]:
### ENTER CODE HERE ###
scoobydoo_merged_year = scoobydoo_merged.copy()
##convert to datetime
#scoobydoo_merged_year['date_aired'] = pd.to_datetime(scoobydoo_merged_year['date_aired'])
##grab year
#scoobydoo_merged_year['date_aired'] = scoobydoo_merged_year['date_aired'].dt.year.astype(int)
##format
#scoobydoo_merged_year['date_aired'] = scoobydoo_merged_year['date_aired'].dt.strftime('%-m/%-d/%Y')
scoobydoo_merged_year['year_aired'] = pd.to_datetime(scoobydoo_merged_year['date_aired'],
                                                     errors='coerce').dt.year.astype('int')
scoobydoo_merged_year['date_aired']

0       9/13/1969
1       9/20/1969
2       9/27/1969
3       10/4/1969
4      10/11/1969
          ...    
598     10/1/2020
599     10/6/2020
600     2/23/2020
601     2/25/2021
602    11/13/2020
Name: date_aired, Length: 603, dtype: object

**Code Check:** Select the `date_aired` and `year_aired` columns from the `scoobydoo_merged_year` DataFrame. The output should match the following for the first and last five rows:

```
date_aired  year_aired
0     9/13/1969        1969
1     9/20/1969        1969
2     9/27/1969        1969
3     10/4/1969        1969
4    10/11/1969        1969
..          ...         ...
598   10/1/2020        2020
599   10/6/2020        2020
600   2/23/2020        2020
601   2/25/2021        2021
602  11/13/2020        2020
```

In [680]:
### CODE CHECK ###
scoobydoo_merged_year[['date_aired','year_aired']]

,date_aired,year_aired
0,9/13/1969,1969
1,9/20/1969,1969
2,9/27/1969,1969
3,10/4/1969,1969
4,10/11/1969,1969
...,...,...
598,10/1/2020,2020
599,10/6/2020,2020
600,2/23/2020,2020
601,2/25/2021,2021


<hr/>

**Exercise12: Total Number of Episodes Per Year**

Use the `scoobydoo_merged_year` DataFrame to create a Series showing the total number of episodes per year. Ensure the results are in **ascending** year order and save the result as a Series named `Exercise12`. 

In [682]:
### ENTER CODE HERE ###
#scoobydoo_merged_year.groupby('year_aired').size().sort_index(ascending=True)

Exercise12 = scoobydoo_merged_year['year_aired'].value_counts().sort_index()

<hr/>

**Exercise13: Total Number of Episodes Per Year and Per Format**

Using the `scoobydoo_merged_year` DataFrame, explore how the total number of episodes varies across years and formats. Organize the data to calculate the total episodes for each unique combination of year and then format. Keep the results in their original order, and assign your output to a Series named `Exercise13`.

In [684]:
### ENTER CODE HERE ###
Exercise13 = scoobydoo_merged_year.groupby(['year_aired', 'format']).size()

<hr/>

**Exercise14: Who Offered the Most Snacks**

According to the creator of the dataset, the `snack_fred`, `snack_daphnie`, `snack_velma`, `snack_shaggy`, and `snack_scooby` columns refers to "when a member of the gang offers someone a Scooby Snack (usually 
given to Shaggy and Scooby). The person they offer does not have to eat it for this to count as TRUE"

Who offers the most snacks? Going back to the `scoobydoo_merged` DataFrame, create a Series that shows the total number of times a scooby snack was offered by each character. Sort the values in descending order, and save the output as a Series named `Exercise14`. You will only include the five columns mentioned above.

In [686]:
### ENTER CODE HERE ###
snack_cols=['snack_fred', 'snack_daphnie', 'snack_velma', 'snack_shaggy','snack_scooby']

Exercise14=scoobydoo_merged[snack_cols].sum().sort_values(ascending=False)

<hr/>

**Exercise 15: Replacing Motive Values**

The dataset creator noted that many entries in the `motive` column have very low counts, with some appearing only once, often in non-traditional series. To clean this column:

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_motives`. Use the new DataFrame for this exercise.
2) Identify the top five most common motives in the `motive` column.
3) Replace any value not in the top five with `Other`.
4) Check the value counts of the updated motive column and save the value counts as `Exercise15`.

In [688]:
### ENTER CODE HERE ###
scoobydoo_motives = scoobydoo_merged.copy()
#scoobydoo_motives['motive'].value_counts()
#motives = ['Competition', 'Theft', 'Treasure', 'Conquer', 'Natural Resource']
#scoobydoo_motives.loc[~scoobydoo_motives['motive'].isin(motives), 'motive'] = 'Other'
#Exercise15 = scoobydoo_motives['motive'].value_counts()

top_motives = scoobydoo_motives['motive'].value_counts().head(5).index.tolist()
scoobydoo_motives.loc[~scoobydoo_motives['motive'].isin(top_motives), 'motive'] = 'Other'
Exercise15 = scoobydoo_motives['motive'].value_counts()

**Code Check:** As a check, after replacing the appropriate values, you should have 188 `Other` entries.

In [690]:
### CODE CHECK ###
Exercise15

motive
Other               188
Competition         168
Theft               125
Treasure             54
Conquer              42
Natural Resource     26
Name: count, dtype: int64

<hr/>

**Exercise16: Binning IMDB Values**

Let's now practice binning the IMDB values. Your task is to put these values into the following categorical bins: `unknown`,`0-4.9`,`5-5.9`,`6-6.9`,`7-7.9`,`8-8.9`,`9-9.9`,`10`. 

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_binning`. Use the new DataFrame for this exercise.
2) The category labels should be ordered and exactly match the above in the exact order.
3) Missing values should be replaced with the `unknown` category.

In [692]:
### ENTER CODE HERE ###
scoobydoo_binning = scoobydoo_merged.copy()
bins = [0,4.9,5.9,6.9,7.9,8.9,9.9,10]
labels = ['0-4.9','5-5.9','6-6.9','7-7.9','8-8.9','9-9.9','10']

scoobydoo_binning['imdb_filled'] = scoobydoo_binning['imdb'].fillna(-1)

scoobydoo_binning['imdb_bins'] =pd.cut(scoobydoo_binning['imdb_filled'], 
                                       bins=bins,
                                       labels=labels,
                                       right=True)
scoobydoo_binning['imdb_bins'] = scoobydoo_binning['imdb_bins'].astype(object)
scoobydoo_binning.loc[scoobydoo_binning['imdb_filled'] ==-1, 'imdb_bins']='unknown'
scoobydoo_binning['imdb_bins'] = pd.Categorical(scoobydoo_binning['imdb_bins'],
                                                categories=['unknown'] + labels, 
                                                ordered=True)

scoobydoo_binning['imdb'] = scoobydoo_binning['imdb_bins']
#scoobydoo_binning['imdb'].cat.categories

<hr/>

**Exercise17: Create Terrain Dummy Values**

Practice creating dummy variables for the `setting_terrain` column.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_terrain`. Use the new DataFrame for this exercise.
2) Encode the `setting_terrain` column into dummy variables, ensuring to drop the first category and set the data type to integers. Use a prefix of `terrain` for the new columns (e.g., `terrain_Urban`, `terrain_Rural`, etc.)

In [694]:
### ENTER CODE HERE ###
scoobydoo_terrain = scoobydoo_merged.copy()
terrain_dummies = pd.get_dummies(scoobydoo_terrain['setting_terrain'], prefix='terrain',
                                 drop_first=False,
                                 dtype=int)
scoobydoo_terrain = pd.concat([scoobydoo_terrain, terrain_dummies], axis=1)

<hr/>

**Exercise18: Catchphrases Per Minute**

For those of you that have ever watched the show, I'm sure you've heard their catchphrases often: 'jeepers', 'jinkies', 'my_glasses', 'just_about_wrapped_up', 'zoinks', 'groovy', 'scooby_doo_where_are_you', 'rooby_rooby_roo'

Analyze which catchphrase is used the most per minute across all episodes. 

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_catchphrase`. Use the new DataFrame for this exercise.
2) In this new DataFrame, calculate a per-minute value for each catchphrase by dividing the respective catchphrase column by the total minutes in the run_time column. Round the results to two decimal places and name the new columns with a `_per_minute` suffix (e.g., `jeepers_per_minute`, `jinkies_per_minute`, etc.).
3) Once the new columns are created, select only these columns, sum their values, and sort the results in descending order. Save the final output as a Series named `Exercise18`.  

In [696]:
### ENTER CODE HERE ###
scoobydoo_catchphrase = scoobydoo_merged.copy()
scoobydoo_merged.head()

catchphrase_cols = ['jeepers', 'jinkies', 'my_glasses', 'just_about_wrapped_up',
                    'zoinks', 'groovy', 'scooby_doo_where_are_you', 'rooby_rooby_roo']
for col in catchphrase_cols: 
    new_col = f"{col}_per_minute"
    scoobydoo_catchphrase[new_col] = (scoobydoo_catchphrase[col] / scoobydoo_catchphrase['run_time']).round(2)

per_minute_cols = [f"{col}_per_minute" for col in catchphrase_cols]

Exercise18 = scoobydoo_catchphrase[per_minute_cols].sum().sort_values(ascending=False)

**Code Check:** As a check, `zoinks_per_minute` should be 66.98.

In [698]:
### CODE CHECK ###
scoobydoo_catchphrase['zoinks_per_minute'].sum()

66.98

<hr/>

**Exercise19: Working with Strings**

If you take a look at the `monster_name`, `monster_gender`, `monster_type`, `monster_subtype`, and the `monster_species` columns, you will notice that some columns have multiple values separated by commas when more than one monster is in the episode. Your goal is to separate these values so that each monster and its corresponding attributes are placed in their own new columns.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_strings`. Use the new DataFrame for this exercise.
2) For each of the monster columns mentioned above, split the values into separate columns. Name the new columns using the pattern `<original_column_name>_part_#`, where # represents the sequence number of the monster (e.g., `monster_name_part_1`, `monster_name_part_2`, etc.). Ensure that the corresponding attributes (e.g., gender, type) follow the same pattern (e.g., `monster_gender_part_1`, `monster_gender_part_2`, etc.).
3) Make sure that you strip any leading or trailing spaces for the values.
    - Run the following code check to ensure the trailing space before "Disguised" is removed. You should see four occurrences of "Disguised": `scoobydoo_strings['monster_type_part_11'].value_counts()`

For example, let's say that an instance of `monster_name` contained three values: `Black Knight`, `Werewolf`, `Dracula`. `Black Knight` would end up in a new column called `monster_name_part_1`. `Werewolf` would end up in a new column called `monster_name_part_2`. `Dracula` would end up in a new column called `monster_name_part_3`. Their respective monster attributes would work similar such as `monster_gender_part_1` for the `Black Knight`'s gender, `monster_gender_part_2` for the `Werewolf`'s gender, `monster_gender_part_3` for `Dracula`'s gender, etc.

Hint: There are many ways to approach this problem. One approach is to use nested loops. The outer loop can iterate through the list of monster columns, and the inner loop can split the values and create the new columns with the appropriate suffixes.

In [701]:
### ENTER CODE HERE ###
scoobydoo_strings = scoobydoo_merged.copy()
monster_cols = ['monster_name', 'monster_gender', 'monster_type', 'monster_subtype', 
                'monster_species']

for col in monster_cols: 
    split_lists = scoobydoo_strings[col].fillna("").apply(lambda x: [s.strip() for
                                                                     s in x.split(',')])

max_parts = split_lists.apply(len).max()

for i in range(max_parts):
    scoobydoo_strings[f"{col}_part_{i+1}"] = split_lists.apply(lambda parts: parts[i] if i < len(parts) else"")
scoobydoo_strings['monster_species_part_11'].value_counts()
#scoobydoo_strings['monster_type_part_11'].value_counts()



monster_species_part_11
         592
Human     11
Name: count, dtype: int64

<hr/>

**Exercise20: Pivot Tables**

Create a pivot table to analyze the average IMDB scores, broken down by network for the rows and format for the columns. Ensure that missing values are filled with `0` in the resulting pivot table, and include both row and column totals in your pivot table. Save the result as `Exercise20`.

In [ ]:
### ENTER CODE HERE ###

Exercise20 = pd.pivot_table(scoobydoo_merged, index='network', columns='format', values='imdb',
               aggfunc='mean', fill_value=0, margins=True, margins_name='All')

Congrats on completing the Scooby-Doo data assignment! You've unmasked data mysteries and earned your spot in the Scooby gang as a Numpy and Pandas pro. Keep solving those data mysteries! 